marimo/_save/loaders/memory.py

```diff
@@ -48,27 +64,35 @@ def load_cache(self, hashed_context: str, cache_type: CacheType) -> Cache:
         key = self.build_path(hashed_context, cache_type)
         if self.is_lru:
             assert isinstance(self._cache, OrderedDict)
-            self._cache.move_to_end(key)
+            assert self._cache_lock is not None
+            with self._cache_lock:
+                self._cache.move_to_end(key)
         return self._cache[key]
 
     def save_cache(self, cache: Cache) -> None:
         key = self.build_path(cache.hash, cache.cache_type)
-        self._cache[key] = cache
         # LRU
         if self.is_lru:
             assert isinstance(self._cache, OrderedDict)
-            self._cache.move_to_end(key)
-            if len(self._cache) > self.max_size:
-                self._cache.popitem(last=False)
+            assert self._cache_lock is not None
+            with self._cache_lock:
+                self._cache[key] = cache
+                self._cache.move_to_end(key)
+                if len(self._cache) > self.max_size:
+                    self._cache.popitem(last=False)
+        self._cache[key] = cache
 
     def resize(self, max_size: int) -> None:
         if not self.is_lru:
             self.is_lru = max_size > 0
             if self.is_lru:
                 self._cache = OrderedDict(self._cache.items())
+                self._cache_lock = threading.Lock()
             self.max_size = max_size
             return
         assert isinstance(self._cache, OrderedDict)
-        while len(self._cache) > max_size:
-            self._cache.popitem(last=False)
+        assert self._cache_lock is not None
```

Comment:
Noticed this after I pushed: Ordered -> dict case isn't preserved (which does matter because the loader is not bound to be flushed)